internal/cose.go

```diff
@@ -0,0 +1,165 @@
+package internal
+
+import (
+	"crypto/ecdsa"
+	"crypto/sha512"
+	"fmt"
+	"math/big"
+
+	"github.com/fxamacker/cbor/v2"
+)
+
+type SigningAlgorithm int
+
+const (
+	BadOrMissingAlgorithm SigningAlgorithm = iota
+	ES384
+)
+
+// https://datatracker.ietf.org/doc/html/rfc8152#section-4.2
+type CoseSign1 struct {
+	_ struct{} `cbor:",toarray"`
+
+	Protected   []byte
+	Unprotected cbor.RawMessage
+	Payload     []byte
+	Signature   []byte
+}
+
+type CoseHeader struct {
+	Alg interface{} `cbor:"1,keyasint,omitempty" json:"alg,omitempty"`
+}
+
+func MakeCoseSign1FromBytes(bytes []byte) (CoseSign1, error) {
+	coseSign1 := CoseSign1{}
+	err := cbor.Unmarshal(bytes, &coseSign1)
+	if nil != err {
+		return CoseSign1{}, fmt.Errorf("unmarshaling CoseSign1 from bytes: %w", err)
+	}
+
+	if len(coseSign1.Protected) == 0 {
+		return CoseSign1{}, fmt.Errorf("missing cose protected headers")
+	}
+
+	if len(coseSign1.Payload) == 0 {
+		return CoseSign1{}, fmt.Errorf("missing cose payload")
+	}
+
+	if len(coseSign1.Signature) == 0 {
+		return CoseSign1{}, fmt.Errorf("missing cose signature")
+	}
+	return coseSign1, nil
+}
+
+func (c CoseSign1) CheckSignature(
+	publicKey *ecdsa.PublicKey,
+) ([]byte, error) {
+	signingAlg, err := c.GetSigningAlgorithm()
+	if err != nil {
+		return nil, fmt.Errorf("extracting signing algorithm: %w", err)
+	}
+
+	// https://datatracker.ietf.org/doc/html/rfc8152#section-4.4
+	sigStruct := struct {
+		_ struct{} `cbor:",toarray"`
+
+		Context     string
+		Protected   []byte
+		ExternalAAD []byte
+		Payload     []byte
+	}{
+		Context:     "Signature1",
+		Protected:   c.Protected,
+		ExternalAAD: []byte{},
+		Payload:     c.Payload,
+	}
```

Comment:
Made `coseSignature` anonymous since it is only used here. Additionally, added link to the section of RFC that defines this structure and how to use it to verify COSE signature